src/groupeddataframe/grouping.jl

```diff
@@ -412,28 +412,53 @@ true
 """
 Base.keys(gd::GroupedDataFrame) = GroupKeys(gd)
 
-# Index with GroupKey
-function Base.getindex(gd::GroupedDataFrame, key::GroupKey)
-    gd === parent(key) && return gd[getfield(key, :idx)]
+
+# The allowed key types for dictionary-like indexing
+const AbstractGroupKey = Union{GroupKey, Tuple, NamedTuple}
+
+
+# Find the integer index of a group given any supported scalar index type
+_findgroup(gd::GroupedDataFrame, i::Integer) = i
+
+function _findgroup(gd::GroupedDataFrame, key::GroupKey)
+    gd === parent(key) && return getfield(key, :idx)
     throw(ErrorException("Cannot use a GroupKey to index a GroupedDataFrame other than the one it was derived from."))
 end
 
-# Index with tuple
-function Base.getindex(gd::GroupedDataFrame, key::Tuple)
+function _findgroup(gd::GroupedDataFrame, key::Tuple)
     for i in 1:length(gd)
-        isequal(Tuple(_groupvalues(gd, i)), key) && return gd[i]
+        isequal(Tuple(_groupvalues(gd, i)), key) && return i
     end
     throw(KeyError(key))
 end
 
-# Index with named tuple
-function Base.getindex(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
+function _findgroup(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
     if length(key) != length(gd.cols) || any(n != _names(gd)[c] for (n, c) in zip(N, gd.cols))
         throw(KeyError(key))
     end
-    return gd[Tuple(key)]
+    return _findgroup(gd, values(key))
 end
 
+# Dictionary-like indexing
+Base.getindex(gd::GroupedDataFrame, key::AbstractGroupKey) = gd[_findgroup(gd, key)]
+
+# Array indexing with dictionary-like keys
+Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray{T}) where {T<:AbstractGroupKey} = gd[[_findgroup(gd, k) for k in idxs]]
```

Comment:
1. This should be fixed as it allows mixing different styles of grouping keys (i.e. mixing named tuples, tuples, and `GroupKey` in a single vector).
2. Also here - and in the whole of this file (also in places you do not touch ) for indexing `AbstractArray` should be changed to `AbstractVector` - as we disallow changing the dimensionality of objects we use here.
3. All `getindex` methods for `GroupedDataFrame` should be moved into the same place in the file as otherwise it will be hard to maintain.